### PR TITLE
Add support for asciinema recordings

### LINKED ISSUE #175: support asciinema player?

### PR DESCRIPTION
**Pull request by dasJ**
Just upload a *.cast file, and it will automatically render with the asciinema player.

Fixes #175.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
thanks for the PR!

```
src/bepasty/views/display.py:116:40: E128 continuation line under-indented for visual indent
```
there's a formatting issue (pep8) that makes travis-ci fail.

also, i'ld like this to be based on a xstatic-package. if there is none yet and you don't want to make one yourself, i can make one in the next days.

**Comment by dasJ** [comment]
@ThomasWaldmann Thanks for the review. I fixed the indentation according to flake8.

I can create a xstatic-package but I wouldn't going to maintain it, so you creating the package would probably make a lot more sense. Could you do that? I'll include it, then

**Comment by ThomasWaldmann** [comment]
OK, I'll try to make a package next week. I'll ping back when finished.

**Comment by dasJ** [comment]
Thanks!

**Comment by ThomasWaldmann** [comment]
https://pypi.org/project/XStatic-asciinema-player/ <- please test it!

**Comment by ThomasWaldmann** [comment]
ping!

**Comment by dasJ** [comment]
Sorry for the long delay, I think it should be okay now

File: src/bepasty/utils/upload.py [hunk 1]
```diff
@@ -66,6 +66,7 @@ def filter_type(cls, ct, ct_hint, filename=None):
         Only allow some basic characters and shorten to 50 characters.
         """
         if not ct and filename:
+            mimetypes.add_type('application/x-asciinema-recording', '.cast')
```
Comment (ThomasWaldmann) [inline src/bepasty/utils/upload.py]: this doesn't look like the right place for this.

i suspect filter_type is a frequently called function and one does not need to call add_type more than once per process.

File: src/bepasty/views/display.py [hunk 2]
```diff
@@ -110,6 +110,10 @@ def read_data(item):
                     src = url_for('bepasty.inline', name=name)
                     link_txt = u'Click to see PDF'
                     rendered_content = Markup(u'<a href="%s">%s</a>' % (src, link_txt))
+                elif ct == 'application/x-asciinema-recording':
+                    src = url_for('bepasty.download', name=name)
+                    rendered_content = Markup(u'<asciinema-player src="%s" title="%s" theme="%s" autoplay>' %
+                                              (src, item.meta[FILENAME], current_app.config['ASCIINEMA_THEME']))
```
Comment (ThomasWaldmann) [inline src/bepasty/views/display.py]: can this be used for XSS by using strange filenames?

Comment (dasJ) [reply src/bepasty/views/display.py]: I doubt it. In my tests, it used the generated filename: `/LdigCtA9/+download`

Comment (dasJ) [reply src/bepasty/views/display.py]: Looks like bepasty handles that. I uploaded `demo.cast" lal.cast` which resulted in:
```
<asciinema-player src="/hFFgG6eb/+download" title="demo.cast lal.cast" theme="asciinema" autoplay="">
```

File: src/bepasty/utils/upload.py (hunk 1 above)
Comment (dasJ) [reply src/bepasty/utils/upload.py]: Yes, I can do that in cli

**Comment by dasJ** [comment]
@ThomasWaldmann Fixed the location of `add_type`, hopefully this is better

File: src/bepasty/cli/server.py [hunk 3]
```diff
@@ -16,6 +17,9 @@ def main():
 
     args = argparser.parse_args()
     app = create_app()
+
+    mimetypes.add_type('application/x-asciinema-recording', '.cast')
```
Comment (ThomasWaldmann) [inline src/bepasty/cli/server.py]: better, but still not the right place.

i guess this gets only invoked when starting the built-in server from flask, but not when using this with an external server (gunicorn, uswsgi, mod-wsgi) just as a wsgi app.

File: src/bepasty/views/display.py (hunk 2 above)
Comment (ThomasWaldmann) [reply src/bepasty/views/display.py]: hmm, ok, so it seems the filename given by the user gets preprocessed somewhere else.

File: src/bepasty/cli/server.py (hunk 3 above)
Comment (ThomasWaldmann) [reply src/bepasty/cli/server.py:21]: did you see my comment?

File: src/bepasty/views/display.py (hunk 2 above)
Comment (ThomasWaldmann) [reply src/bepasty/views/display.py:116]: guess this should have a theme default like:
```
current_app.config.get('ASCIINEMA_THEME', 'asciinema')
```

Comment (ThomasWaldmann) [reply src/bepasty/views/display.py:116]: bepasty preprocesses the filename before storing it into metadata and uses this regex (all matching chars will be stripped from filename):

```
class Upload(object):
    _filename_re = re.compile(r'[^a-zA-Z0-9 \*+:;.,_-]+')
```

File: src/bepasty/cli/server.py (hunk 3 above)
Comment (ThomasWaldmann) [reply src/bepasty/cli/server.py:21]: check if this should be on top level of module, where mimetypes is first imported.
maybe right below the import statement.

**Comment by ThomasWaldmann** [comment]
Note: i cleaned up some comments that are not relevant any more.

Stuff left is about TODO and about "no XSS here".

**Comment by ThomasWaldmann** [comment]
@dasJ do you want to finish this or shall I?

File: src/bepasty/views/display.py (hunk 2 above)
Comment (ThomasWaldmann) [reply src/bepasty/views/display.py:116]: i fixed that myself.

File: src/bepasty/cli/server.py (hunk 3 above)
Comment (ThomasWaldmann) [reply src/bepasty/cli/server.py:21]: fixed this myself.

**Comment by ThomasWaldmann** [comment]
Thanks for the PR! :-)

**Pull request merged**